### PR TITLE
Connect: Implement SetGatewayLocalPort RPC

### PR DESCRIPTION
**Pull request by ravicious**
We want the user to be able to change the randomly assigned port number of a gateway (gravitational/webapps.e#302).

To accomplish this, we need an RPC that's going to change the port number and restart the gateway under a new port. That should happen only if the new port is actually free – if not then the RPC should fail and the gateway should continue running under the old port.

The PR might seem bigger than it actually is due to protobuf changes which are encapsulated within a single commit. Best reviewed commit-by-commit.

---

* Why not make an RPC called `UpdateGateway` instead of having both `SetGatewayTargetSubresourceName` and `SetGatewayLocalPort`?
   * Initially we had a slightly different idea on how those two RPCs are going to behave than it turned out in practice. Now it sure seems like they could be merged into a single `UpdateGateway` RPC.
   * Honestly, this occured to me only after I finished everything and started to create this PR. At this point I'd rather spend the time on delivering other features and IMHO we can postpone this refactor until we need to update a third field on the gateway (if ever).
* How do I test this by myself?
   * There's a `connect/db-port-wip` branch in the webapps repo. You can build Connect off that branch and after connecting to the db you should see a form field for changing the port number. Changing the value makes calls this new handler.

---

A small reminder that [lib/teleterm has a readme](https://github.com/gravitational/teleport/tree/ravicious/gateway-hash-map/lib/teleterm) in case this is your first foray into this directory.

### CODE REVIEW

File: lib/teleterm/clusters/dbcmd_cli_command_provider_test.go [hunk 1]
```diff
@@ -83,49 +86,61 @@ func TestDbcmdCLICommandProviderGetCommand(t *testing.T) {
...
+					CLICommandProvider:    dbcmdCLICommandProvider,
 				},
-			}
-			fakeStorage := fakeStorage{
-				clusters: []*Cluster{&cluster},
-			}
-			dbcmdCLICommandProvider := NewDbcmdCLICommandProvider(fakeStorage, fakeExec{})
+			)
+			require.NoError(t, err)
+			t.Cleanup(func() { gateway.Close() })
```
Comment (ravicious) [inline lib/teleterm/clusters/dbcmd_cli_command_provider_test.go:109]: Since `gateway.Config` became a private field and is no longer embedded within `gateway.Gateway`, I had to create a gateway here through `gateway.New`.

File: lib/teleterm/gateway/gateway.go [hunk 2]
```diff
@@ -148,14 +193,12 @@ func (g *Gateway) CLICommand() (string, error) {
 
 // Gateway describes local proxy that creates a gateway to the remote Teleport resource.
 type Gateway struct {
-	Config
-
+	cfg        *Config
```
Comment (ravicious) [inline lib/teleterm/gateway/gateway.go:253]: Previously any caller could change any public field on `Config` which would actually be detrimental to the main subject of this PR. We don't want people to be able to change the underlying `LocalPort` without actually restarting the gateway.

This enables us to encapsulate this logic within a setter method rather than depending on the promise that we'll remember to always restart the gateway after changing `gateway.LocalPort`.

File: lib/teleterm/gateway/gateway.go [hunk 3]
```diff
@@ -32,19 +32,46 @@ import (
...
 func New(cfg Config) (*Gateway, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%s", cfg.LocalAddress, cfg.LocalPort))
+	result, err := newListenerAndLocalProxy(cfg, cfg.LocalPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
Comment (ravicious) [inline lib/teleterm/gateway/gateway.go:45]: Most of the functionality of `New` and `SetLocalPort` is almost identical – they both need to create a listener on the given port and then initialize a local proxy with that listener.

The difference is that `New` then just returns a new instance of the `Gateway` struct while `SetLocalPort` mutates an existing instance, as well as stops the previous proxy and starts a new one.

I extracted the shared logic into `newListenerAndLocalProxy`.

File: lib/teleterm/gateway/gateway.go [hunk 4]
```diff
@@ -32,19 +32,46 @@ import (
...
+
+type newListenerAndLocalProxyResult struct {
+	LocalPort    string
+	listener     net.Listener
+	closeContext context.Context
+	closeCancel  context.CancelFunc
+	localProxy   *alpn.LocalProxy
+}
+
+func newListenerAndLocalProxy(cfg Config, port string) (*newListenerAndLocalProxyResult, error) {
```
Comment (ravicious) [inline lib/teleterm/gateway/gateway.go:67]: I'm actually not sure what's the common approach in Go when I need to return more than two values from a method. I opted for creating this struct specifically for this purpose but I'm open to suggestions.

File: lib/teleterm/gateway/gateway.go [hunk 5]
```diff
@@ -58,11 +85,21 @@ func New(cfg Config) (*Gateway, error) {
 	}()
```
Comment (ravicious) [inline lib/teleterm/gateway/gateway.go:85]: Now this is interesting because if you look at the previous version of `gateway.New`, [this deferred function was "guarding" almost the whole `New` function](https://github.com/gravitational/teleport/blob/a41a4300708c3a72f2ed58ae9f1eeec89f5dcd87/lib/teleterm/gateway/gateway.go#L47-L56). So if `gateway.New` failed at any point after creating the listener, the context would be canceled and the listener closed.

In this version, this deferred function guards only `newListenerAndLocalProxy`. I think this is fine because both `New` and `SetLocalPort` don't currently have any calls that might fail after they call `newListenerAndLocalProxy`.

It might be tricky to keep this as invariant going forward though. OTOH, I had no idea how to write this and employing some callbacks/channels for this seemed too complicated for me here, though I might be wrong.

File: lib/teleterm/gateway/gateway.go (hunk 4 above)
Comment (ravicious) [inline lib/teleterm/gateway/gateway.go:67]: I made it so that functions which don't mutate `Config` accept a value rather than a pointer.

File: lib/teleterm/gateway/gateway.go [hunk 6]
```diff
@@ -172,6 +196,39 @@ func (g *Gateway) LocalPort() string {
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	previousPort := g.cfg.LocalPort
+
+	g.closeCancel()
+	if err = g.localProxy.Close(); err != nil {
+		g.cfg.Log.WithError(err).Warnf("Failed to close the previous proxy on port %s.", previousPort)
+	}
```
Comment (ravicious) [inline lib/teleterm/gateway/gateway.go:216]: Is failing to close a listener a real concern usually?

Of course the logic in `alpn.LocalProxy.Close()` might grow and include other stuff than just closing the listener. OTOH if `SetLocalPort` returned an error in this scenario, I guess we'd need to close the new listener then. Maybe that would be a behavior that could be described as more correct?

File: lib/teleterm/gateway/gateway_test.go [hunk 7]
```diff
@@ -92,6 +89,164 @@ func TestGatewayStart(t *testing.T) {
...
+	// Verify that the old context was canceled.
+	//
+	// What we really want to test is if the old listener was closed. Unfortunately, we don't seem to
+	// have a straightforward way to test this as at this point another process might have started
+	// listening on that port.
+	require.ErrorIs(t, originalCloseContext.Err(), context.Canceled,
+		"The listener on the old port wasn't closed after starting a listener on the new port.")
+}
+
+func TestSetLocalPortDoesntStopGatewayIfNewPortIsOccupied(t *testing.T) {
```
Comment (ravicious) [inline lib/teleterm/gateway/gateway_test.go:119]: I could make a table driven test out of those two but is it worthwhile doing this for just two tests?

As someone who started writing Go only after joining Teleport I'm super biased but most of the time the table driven tests in our codebase are far less readable to me than if they were just regular tests with logic shared through regular functions.

File: lib/teleterm/gateway/gateway.go (hunk 3 above)
Comment (ravicious) [reply lib/teleterm/gateway/gateway.go]: Hmm, good question. After thinking about it for a while I realized this: `RestartGateway` is a workaround which we're going to remove as soon as we implement gravitational/webapps.e#299. `gateway.Gateway` itself doesn't know how to refresh certs so it had to be done on the daemon level.

We could make `daemon.Service.SetGatewayLocalPort` create a new gateway under a different port and remove the old one but we'd need to pick either one of these three options:

1. Expose almost all fields on `gateway.Gateway` so that `daemon.Service` can copy them over. It'd need to do the equivalent of this but take all the values from the old gateway:

https://github.com/gravitational/teleport/blob/02b4f8575f5f0bc305a37c89b58e4469c2662921/lib/teleterm/clusters/cluster_gateways.go#L51-L63

2. Make `SetGatewayLocalPort` pull all the initialization data that `clusters.Cluster.CreateGateway` pulls (as shown above) without fetching new certs but then we'd still be making unnecessary calls to the cluster, for example to get the db so that we can then call `db.GetName()`.

https://github.com/gravitational/teleport/blob/02b4f8575f5f0bc305a37c89b58e4469c2662921/lib/teleterm/clusters/cluster_gateways.go#L42

3. Do a mix of 1 and 2, exposing some methods but not all just so that we can avoid those unnecessary calls.

In the end, currently `gateway.SetLocalPort` does only the absolutely necessary work to carry out the port change.

The combination of `RestartGateway` being a workaround and letting the struct manipulate its own internals makes me lean towards keeping `SetLocalPort` as it is.

What do you think?

Comment (greedy52) [reply lib/teleterm/gateway/gateway.go]: It's just I saw the `RestartGateway` function and wondered about that approach.

> The combination of RestartGateway being a workaround and letting the struct manipulate its own internals makes me lean towards keeping SetLocalPort as it is.

Yes, I think abstracting things inside the gateway like `SetLocalPort` is better with your explanations. We just need to make sure the local proxy swap inside the gateway is done properly. Let me do a detailed review 😄

Comment (ravicious) [reply lib/teleterm/gateway/gateway.go]: It's also kinda my fault for not documenting it better that `RestartGateway` is only a workaround – I mentioned it only in `lib/teleterm/api/proto/v1/service.proto` but people are less likely to look there than at the actual implementation.

File: lib/teleterm/gateway/config.go [hunk 8]
```diff
@@ -83,5 +88,13 @@ func (c *Config) CheckAndSetDefaults() error {
 		return trace.BadParameter("missing target URI")
 	}
 
+	if c.CLICommandProvider == nil {
+		return trace.BadParameter("missing CLICommandProvider")
+	}
+
+	if c.TCPPortAllocator == nil {
+		return trace.BadParameter("missing TCPPortAllocator")
```
Comment (greedy52) [inline lib/teleterm/gateway/config.go:96]: should `NetTCPPortAllocator` be the default?

File: lib/teleterm/gateway/gateway.go [hunk 9]
```diff
@@ -116,13 +141,90 @@ func (g *Gateway) Close() error {
...
+	g.cfg.LocalPort = result.LocalPort
+	g.localProxy = result.localProxy
+	g.closeCancel = result.closeCancel
+	g.closeContext = result.closeContext
+
+	go func() {
+		if err := g.Serve(); err != nil {
+			g.cfg.Log.WithError(err).Warn("Failed to restart the gateway after changing the port.")
+		}
+	}()
```
[the remark this answers is not in the document]
Comment (ravicious) [reply lib/teleterm/gateway/gateway.go:227]: Any operation on gateways must go through `daemon.Service` methods which obtain a write lock before making any modifications, so I think we're protected against that. Well, unless we write code within `gateway` which calls those functions twice.

If `g.Serve` is never called before `SetLocalPort` then I think everything should be fine? The context will be closed which will close the listener and then `Close` on local proxy will be a noop I think.

Is it worthwhile to add another lock on the `gateway.Gateway` layer that would be called in setters?

File: lib/teleterm/gateway/config.go (hunk 8 above)
Comment (ravicious) [reply lib/teleterm/gateway/config.go:96]: Ah, sure, I can make it so. I wasn't sure if it's a common approach to define defaults for those "collaborator" objects or if I should explicitly request them.

Comment (greedy52) [reply lib/teleterm/gateway/config.go:96]: Not a big deal either way. I think the question is if we see other "production" code will be using a different `TCPPortAllocator`. If so, current way is the best. If not, and the only purpose is overwriting for testing, might set a default if not provided (also in that case you can keep `tcpPortAllocator` private.)

File: lib/teleterm/gateway/gateway.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/teleterm/gateway/gateway.go:227]: > Any operation on gateways must go through daemon.Service methods which obtain a write lock before making any modifications

Ah, I missed the lock part on the outside. Maybe let's add a comment somewhere how they are protected.

> If g.Serve is never called before SetLocalPort then I think everything should be fine? The context will be closed which will close the listener and then Close on local proxy will be a noop I think.

Yes, closing the not-started local proxy should be fine. Though, I am not sure if there is any side effect if `g.Serve` is called twice on the same proxy. Probably ok too. But more importantly, if `SetLocalPort` is called first, the caller may not expect the local proxy to start until the caller decides to call `Serve`. If there is no easy way to address this, maybe leave a comment `SetLocalPort` is expected to be called after `Serve`. What do you want?

File: lib/teleterm/gateway/config.go (hunk 8 above)
Comment (ravicious) [reply lib/teleterm/gateway/config.go:96]: I'll make it the default then.

File: lib/teleterm/gateway/gateway.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply lib/teleterm/gateway/gateway.go:227]: > Ah, I missed the lock part on the outside. Maybe let's add a comment somewhere how they are protected.

Sure.

> But more importantly, if `SetLocalPort` is called first, the caller may not expect the local proxy to start until the caller decides to call `Serve`. If there is no easy way to address this, maybe leave a comment `SetLocalPort` is expected to be called after `Serve`. What do you think?

I suppose one way to solve this would be to add a private field called sth like `serving` and then not do the equivalent of `Serve` if `serving` is true. But even then what would irk me the most is that `SetLocalPort` pretends to be a simple setter. In reality, there's more logic involved which just has to be contained somewhere.

I think for now I'll rename `SetLocalPort` to `SetLocalPortAndRestart` and add a comment saying that `SetLocalPortAndRestart` is expected to be called after `Serve`.

File: lib/teleterm/gateway/gateway.go [hunk 10]
```diff
@@ -56,11 +85,21 @@ func New(cfg Config, cliCommandProvider CLICommandProvider) (*Gateway, error) {
...
+		return nil, trace.Wrap(err)
+	}
+
+	localProxy, err := newLocalProxy(closeContext, cfg, listener)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
+	ok = true
+	return &newListenerAndLocalProxyResult{localPort, listener, closeContext, closeCancel, localProxy}, nil
```
Comment (greedy52) [inline lib/teleterm/gateway/gateway.go:99]: nitpick: use field names to initialize.

File: lib/teleterm/gateway/gateway.go (hunk 4 above)
Comment (greedy52) [reply lib/teleterm/gateway/gateway.go:67]: I think normally I would make this `Gateway.init` function to assign directly but I understand it will not work for the swap case.

Not sure if letting `Gateway` own this new struct will be better than holding each piece separately. something like:
```
type Gateway struct {
	cfg   *Config
        proxy *gatewayProxy
}
type gatewayProxy struct {
        *alpn.LocalProxy
	localPort    string
	closeContext context.Context
	closeCancel  context.CancelFunc
}
```

**Review by greedy52 — APPROVED** [review]

File: lib/teleterm/gateway/gateway.go (hunk 4 above)
Comment (ravicious) [reply lib/teleterm/gateway/gateway.go]: The only downside to that solution would be that we'd have `Config.LocalPort` and `gatewayProxy.localPort`. I think I'd rather keep those fields as they are right now so that we can avoid having do deal with that duplication.

**Comment by ravicious** [comment]
@nklaassen Could you take a look at this PR?

File: lib/teleterm/gateway/gateway.go (hunk 4 above)
Comment (greedy52) [reply lib/teleterm/gateway/gateway.go]: IMHO, it's better to have them separate. `Config.LocalPort` is the static input and shouldn't be overwritten. `gatewayProxy.localPort` is the real deal and is what should be returned by `gateway.LocalPort`. But that's just me. I am Ok with current approach. Let's if others have other suggestions.

File: lib/teleterm/gateway/gateway.go [hunk 11]
```diff
@@ -116,29 +145,114 @@ func (g *Gateway) Close() error {
...
+	}
+
+	g.cfg.LocalPort = result.LocalPort
+	g.localProxy = result.localProxy
+	g.closeCancel = result.closeCancel
+	g.closeContext = result.closeContext
+
+	go func() {
+		if err := g.Serve(); err != nil {
+			g.cfg.Log.WithError(err).Warn("Failed to restart the gateway after changing the port.")
```
Comment (nklaassen) [inline lib/teleterm/gateway/gateway.go:237]: Is it a really a failure to restart, or could it be any other error that happens later on?

File: lib/teleterm/gateway/gateway.go [hunk 12]
```diff
@@ -116,29 +145,114 @@ func (g *Gateway) Close() error {
...
+func (g *Gateway) Log() *logrus.Entry {
+	return g.cfg.Log
+}
+
+func (g *Gateway) LocalAddress() string {
+	return g.cfg.LocalAddress
+}
+
+func (g *Gateway) LocalPort() string {
+	return g.cfg.LocalPort
```
Comment (nklaassen) [inline lib/teleterm/gateway/gateway.go:200]: Could it be a race?

File: lib/teleterm/gateway/gateway.go [hunk 13]
```diff
@@ -116,29 +145,114 @@ func (g *Gateway) Close() error {
...
+		g.cfg.Log.WithError(err).Warnf("Failed to close the previous proxy on port %s.", previousPort)
+	}
+
+	g.cfg.LocalPort = result.LocalPort
+	g.localProxy = result.localProxy
+	g.closeCancel = result.closeCancel
+	g.closeContext = result.closeContext
+
+	go func() {
+		if err := g.Serve(); err != nil {
```
Comment (nklaassen) [inline lib/teleterm/gateway/gateway.go:236]: it makes me uneasy that we just fire a `Serve()` and never wait for it to finish or consume the error, probably outside the scope of this PR but I think it's going to be difficult to gracefully handle errors and restarts. In teleport we have `RegisterCriticalFunc` so that if things like this fail, we can restart. This is why I also think it's not great that we let the original `Serve()` goroutine die and just transparently call a new one here, in case we ever want to have some error handling for these in teleterm.

**Review by nklaassen — COMMENTED** [review]
I was going to make the same comment about creating a brand new gateway rather than mutating the one that's already established, but I see that's already been discussed. Fwiw, I don't think the extra `GetDatabase` call would be that bad if we could greatly simplify the logic by just calling the same function the create a gateway every time. The concurrency with `Serve()` and getting/setting values without locks, different contexts that may change, it's a bit hard to follow now. If you think it's feasible I would do it, otherwise I just have a few comments on the current implementation.

File: lib/teleterm/gateway/gateway.go [hunk 14]
```diff
@@ -116,29 +145,114 @@ func (g *Gateway) Close() error {
...
+	}
+
+	result, err := newListenerAndLocalProxy(*g.cfg, port)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	previousPort := g.cfg.LocalPort
+
+	g.closeCancel()
```
[the remark this answers is not in the document]
Comment (ravicious) [reply lib/teleterm/gateway/gateway.go:225]: > I don't think you need to call this, just closing the localProxy should be fine (this will be relevant if you use a single parent context, you won't be able to cancel it here)

Yeah, I noticed that it might be unnecessary too but so far I was just copying this around without thinking. I'll take care of that in the new version.

File: lib/teleterm/gateway/gateway.go (hunk 13 above)
Comment (ravicious) [reply lib/teleterm/gateway/gateway.go:236]: > This is why I also think it's not great that we let the original `Serve()` goroutine die and just transparently call a new one here, in case we ever want to have some error handling for these in teleterm.

How should this look like if I wanted to attempt to improve it in the future? Should there be a dedicated goroutine that exclusively manages the lifecycle of the gateway and restarts it on failure?

**Comment by ravicious** [comment]
@nklaassen @greedy52 Re: creating a new gateway rather than mutating an existing one: since you both seem to favor that approach and given that I cannot really argue with it being simpler, I decided to start work on a new implementation. It'll have a function called sth like `gateway.NewWithLocalPort`. It'll accept an existing gateway and the port and then it'll basically call `New` by passing the same config minus the port which will be different.

I'll make a new PR for that just to not complicate this one further.

I really want to avoid any calls to cluster in `daemon.SetGatewayLocalPort` because having just one means having to handle expired certs and so on. Not calling the cluster means that the SetGatewayLocalPort RPC is going to be a simple local call done entirely on the client side and I'd like to keep it that way.
